### PR TITLE
Associations Project: Private Events

### PR DESCRIPTION
**Pull request by Leandro-Barretoo**
In this project, the associations were the main goal to achieve and contains:
many-to-one between User(creator) and Event (on the creation of the event)
many-to-many between User(attendee) and Event(attended_event)

ps: if you run `rspec` to run RSpec and didn't work, try bundle exec `rspec`

### CODE REVIEW

**Review by NtwaliHeritier — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUESTED🛑🛑**

Hello team,
Great work. All I can say is congrats. Great implementation of the project and it is professional looking.

Let's make some few changes to get it approved.

[REQUIRED]: In the event controller, index method, put two variables to implement the scopes you defined in the event model. e.g: @past_events, @future_events.

[REQUIRED]: The attend button is crashing. Kindly check it out

[REQUIRED]: The logout button is crashing. Kindly fix it.

[REQUIRED]: In the users controller, in the show method, specify variable for showing previous events attended by current_user and events he/she will attend in the future

Kindly make the requested changes and ask for another review👌👍

Happy coding💻💻💻💻

File: app/views/events/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      <li><%= ev.description %><br><%= ev.date %></li><%= link_to "View", event_path(id: ev.id) %><br>
+      <% if logged_in? %>
+        <%= link_to "Attend", event_attendences_path(event_id: ev.id), method: :post %>
+      <% end %>
+    <% end %>
+  </ul>
+</div>
+<% if logged_in? %>
+  <%= link_to "Create Event", new_event_path, class: "text-center" %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/events/index.html.erb]: There should be absolutely NO logic in views.

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -1,16 +1,23 @@
...
 
   <body>
+    <%= render "shared/navb" %>
+    <% if notice %>
+      <%= notice %>
+    <% end %>
+    <% if alert %>
+      <%= alert %>
+    <% end %>
     <%= yield %>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb]: There should be absolutely NO logic in views.

File: app/views/shared/_navb.html.erb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            <%= link_to "Log Out", logout_path, method: :delete, class: "nav-link" %>
+          </li>
+        <% else %>
+          <li class="nav-item">
+            <%= link_to "Login", login_path, class: "nav-link" %>
+          </li>
+          <li class="nav-item">
+            <%= link_to "Sign Up", new_user_path, class: "nav-link" %>
+          </li>
+        <% end %>
```
Comment (acushlakoncept) [inline app/views/shared/_navb.html.erb]: There should be absolutely NO logic in views.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Excellent progress so far.  Both of you have done well on this project.
Let's make a few changes, and we will be done.

- [ ] Let's maintain a RESTful route by ensuring we are only exposing the available routes to our users. Since your `users_controller` only has `new, create and delete` action. 
Set `resources :users, only: [:new, :create, :show]

- [ ] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)

I have gone ahead and made some comments on your codebase:

- [ ] Do not forget to add `yarn` to your getting started instructions, perfect after the `bundle install`. I couldn't run the project without running `yarn`

My personal highlights: I know sometimes it feels like a waste of time to pay attention to front-end design when you are been tested on backend stuff. Constant practice on this little UI does help serve as a consistent practice. Projects become appealing to check out even though it's doing just a little. Always try as much
### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/) as possible to pay attention to your UI, just a little while doing your projects here.

**Comment by Leandro-Barretoo** [comment]
I need the logic in the view. The rest is fixed

**Review by mke2111 — CHANGES_REQUESTED** [review]
# **STATUS: CHANGES REQUIRED 🔴 ⚠️** 

## Hey @Leandro-Barretoo & @wintan1418, 👋🏽 
 **Thank you for the good job you got going on.** Your app works just fine.

## REQUIRED CHANGES

- [ ] Kindly move all the logic (conditional statements and loops) from the view files. Transfer them to the helpers folder. You can go ahead and create methods that you will call in the view files to provide the same functionality.

**Kindly make all the requested changes and submit another CRR. 👍🏽** 

## Cheers!! 🍹 & Happy Coding.

# Reach me on slack for a chitchat 🆘  **Roy Mukuye**

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- Good progress :+1: 
- No inline comment :+1: 
- No logic in your view :100: 

To Improve :factory_worker:

- [ ] It will be nice if you add `.stylelintrc.json` in your project as  it represent style guidelines that you share with your team - which is a group of all Microverse students.
- [ ] It will be good if `create` button is not displayed when user is not signed-in in order to make your app more intuitive:


- [ ] Please add link that leads to the page where user can attend one event because your APP has to display all list of attendants in the event's show page, I think, it will be nice if there is a page that gives user opportunity to attend one event.
- [ ] I got this error when I run your APP, please add this command `yarn install` in your readme file, before `rails server` to make sure that `application.js` is present in your APP


- [ ] Some of your tests don't pass, it means that there are some bugs in your APP please, make sure to fix it








That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Changes Required 🙏🏿 
Nice work making the requested changes. I was able to successfully run the app locally. Well done!
Let's make some changes so we can merge this PR.
- [ ] You have on failing test case: `rspec ./spec/event_spec.rb:73 # Events when a user attends an event`
- [ ] On clicking the `Attend` button, I got this error.
review
Right now a user can attend an event that doesn't exist. This is probably the reason for that error. Make sure a user can attend an event that was created by other users.


### Martin Afani 
#### Slack: whiz25

**Review by SmithVyne — CHANGES_REQUESTED** [review]
### Changes required 🤕 👍 
Hello, most of your project works fine but I can't find a button or link that a user can click to attend an event. 

image


[OPTIONAL] - I think you should add a link to the create new event page on the index page.

Make these changes and request another code review. You should get it approved this time.

You can also reach out to me on slack, **@SmithVyne** if you have any other questions.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @wintan1418  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implement the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] When a user is added to the past event (means list of already attended events), and when they try to browse the `myEvents` , the app crashes with the error on the screenshot 

Screenshot from 2021-03-18 12-13-51

- [ ] Navigation links are present at the top of the page => The new_event and all_events link is missing from the top of the page
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Comment by wintan1418** [comment]
hy Mugirase Emmanuel (descholar),,you have  an idea of how i can get to solve this issue,and also this error you are getting,I checked through and I saw that you are getting it from app/views/shared/table.html.erb,but in my shares directory ,I don't have any table.html.erb

**Comment by descholar-ceo** [comment]
> hy Mugirase Emmanuel (descholar),,you have an idea of how i can get to solve this issue,and also this error you are getting,I checked through and I saw that you are getting it from app/views/shared/table.html.erb,but in my shares directory ,I don't have any table.html.erb

Hello @wintan1418 I do appologise for sending you a wrong screenshot, I have just updated the comment, you can check the screenshot again: Screenshot from 2021-03-18 12-13-51

**Comment by wintan1418** [comment]
i have done the corrections,waiting for a review

**Review by oloomoses — APPROVED** [review]
# Congratulations, Project Approved. :muscle: :tada: :clap: 
**Great job working on the suggested changes from the previous review.
You have passed all the requirements for this project :muscle: 
Feel free to merge and close this pull request.**

**I wish you the best in your next project and see you at the top :rocket:** 
#### Happy Coding
_Slack: Oloo Moses_

**Comment by wintan1418** [comment]
hallelujah,thanks to all you TSE, though frustrating at some point.

**Pull request merged**